### ISSUE TITLE
Rename ApplicationContext.ServiceProviderScope

### ISSUE THREAD

**Issue by ajohnstone-ks** [MarimerLLC/csla#1999]
**Is your feature request related to a problem? Please describe.**
Per this [discussion](https://github.com/MarimerLLC/csla/discussions/1991), rename ApplicationContext.ServiceProviderScope to something more intuitive.

Maybe ApplicationContext.CurrentServiceProvider?

**Comment by dazinator** [comment]
`ApplicationContext` provides access to two different `IServiceProvider`'s - there is the `root` one - which is exposed via the property named `DefaultServiceProvider`, and the scoped one, exposed by the property `ServiceProviderScope` - they can be seen on this commit:

https://github.com/MarimerLLC/csla/pull/2011/commits/cd51d56f05689f2ae5be29b612aff2ad0cd1aff1

You can see however that `ApplicationContext` doesn't actually hold these references itself, it sources them from the correct `ContextManager` at the time - of which there could be two - either the default `ContextManager` or the `WebContextManager`. This means there are currently 4 possible backing fields for these two different IServiceProvider's.

With respect to the `root` IServiceProvider I am not sure of the benefit of having the individual `ContextManager`s hold the reference to the `root`  IServiceProvider. Perhaps this needs to be moved into `ApplicationContext` as a static property (I generally try to avoid static properties though) or perhaps it should only ever be sourced from the default `ContextManager` - basically just want to achieve the goal of having one single backing field for this? It doesn't seem to make a difference which `context` you want the root provider for (Web vs non web) as the root IServiceProvider will always be the same in either case, so having it as it is currently, where its a field on the two different `ContextManager`s probably isn't the best fit?

We should also look at when / why CSLA needs this `root` IServiceProvider. CSLA will usually always want to get the current scoped IServiceProvider, and there will be some CSLA logic that may fallback to the root IServiceProvider if there is no scoped IServiceProvider available. AFAIK this is the only reason CSLA will need a reference to the root IServiceProvider. Therefore it may make sense to restrict access to the root IServiceProvider, by making it an internal property, so external code can set the root IServiceProvider, but can't get at it. This will enforce the idea that code external to the CSLA framework shouldn't be using this mechanism to get access to the root IServiceProvider - (code outside CSLA can grab and manage a reference to the root IServiceProvider for its own purposes, in its own way) - but should only be handing it to CSLA so CSLA can do it its thing.

With respect to naming, the convention used in .net core for the `root` IServiceProvider is to refer to it as `ApplicationServices`.
For example, the `ApplicationServices` property on `IApplicationBuilder` interface: https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.builder.iapplicationbuilder?view=aspnetcore-5.0

Therefore, for the root IServiceProvider, I propose CSLA ApplicationContext has a method like this:

```
ApplicationContext.SetApplicationServices(IServiceProvider sp);

```
and the property for CSLA to `get` this can be internal to CSLA so it can use it as a fallback where necessary.

Moving on to the current / scoped IServiceProvider that CSLA will need to obtain (or lazily create under some setup's, as discusses elsewhere) - asp.net core doesn't really set any decent precedent for what to name this. When asp.net core creates a scoped IServiceProvider for a request, you can access it on `IHttpContext.RequestServices` - so the asp.net core convention seems to be to name the property like this: `[ScopeName]Services` - which gives rise to `ApplicationServices` and `RequestServices`.
However in CSLA's case, it doesn't know (or care) whether the scoped IServiceProvider given to it represents a Request, or some other more granular scope, like a background job, or an individual data portal operation. Therefore its a bit hard to follow the convention `[ScopeName]Services` when the scopes could represent anything. So I am left with the following suggestions:

- ApplicationContext.Services; 
- ApplicationContext.CurrentServices;

**Comment by dazinator** [comment]
Also thanks @ajohnstone-ks for raising this issue. I had avoided renaming this property as part of my previous PR, because it widened the scope of the changes - i.e more noise in terms of lines modified that didn't actually contribute to my objective in that PR - but I absolutely agree changes need to be made.

**Comment by dazinator** [comment]
Thinking more on my comment:

> I am not sure of the benefit of having the individual `ContextManager`s hold the reference to the `root`  IServiceProvider. Perhaps this needs to be moved into `ApplicationContext` as a static property (I generally try to avoid static properties though) or perhaps it should only ever be sourced from the default `ContextManager` - basically I just think the there should one single backing field for the root IServiceProvider, as it doesn't seem to make a difference which `context` you want the root provider in (Web vs non web) as the root IServiceProvider will always be the same in either case - and CSLA would only ever want to use this, when there was no more granular scope available. For example in a console app that doesn't use scopes it might be valid to operate within the root scope only.

I think a static property should be avoided, as that complicates unit testing. Ideally, concurrent CSLA unit tests would be able to run, each thread settings its own root `IServiceProvider` if needed to provide isolation between the different test scenarios. Therefore this would have to be an instance level `AsyncLocal<T>` field, but we just want a single field for CSLA to consume this, not one on ContextManager and another on WebContextManager.

**Comment by ajohnstone-ks** [comment]
I think keeping both the DefaultServiceProvider & CurrentServicePrivder is important; Csla needs to track the Default (or Root, if you'd prefer).  The Default/Root is the one you configure and set once at application startup, just as you create & configure a single root container at start up.   You then don't actually use the Default/Root to resolve anything, its would be for just `GetRequiredService<IServiceScopeFatory>` allowed scopes to be created & disposed as needed; the scopes Csla uses to inject would be the Current, so that disposables are cleaned up properly after the data portal methods finish.

I would also prefer keeping Provider in the name; Services makes me think of a collection, but you're not getting a collection of services, you're getting something that and resolve a service for you, however it does so.  

I'm on the fence about where the Default/Root SP should be kept.  Is the intent that you set the Default/Root once at application startup?  If so it should probably be static, provided you're only using the Default/Root to create new scopes.  The scopes then would isolate the unit tests.

**Comment by ajohnstone-ks** [comment]
@dazinator 

I think this is a discussion for another issue altogether, but actually I think that if the answer to how to manage scopes is "do it via an interceptor," then really both DefaultServiceProvider & CurrentServiceProvider need to be fully public.  

This is actually the code for the interceptor I wrote (not the one that I put in the other ticket hoping to become part of Csla).  Quite a bit of hop-jumping, IMO because I couldn't read those two properties.

```
public sealed class ServiceScopeInterceptor : IInterceptDataPortal {
	private const string ScopeKey = "$$current.scope";

	private static IServiceScopeFactory scopeFactory;
	
	private bool createdScope;

	public static void SetScopeFactory(IServiceScopeFactory scopeFactory) =>
		ServiceScopeInterceptor.scopeFactory = scopeFactory ?? throw new ArgumentNullException(nameof(scopeFactory));

	public void Initialize(InterceptArgs e) {
		if (ApplicationContext.LocalContext[ScopeKey] == null) {
			createdScope = true;

			var scope = scopeFactory.CreateScope();
			ApplicationContext.LocalContext[ScopeKey] = scope;
			ApplicationContext.ServiceProviderScope = scope.ServiceProvider;
		}
	}

	public void Complete(InterceptArgs e) {
		if (createdScope && ApplicationContext.LocalContext[ScopeKey] is IServiceScope scope) {
			ApplicationContext.LocalContext[ScopeKey] = null;
			ApplicationContext.ServiceProviderScope = null;
			scope.Dispose();
		}
	}
}
```

**Comment by dazinator** [comment]
@ajohnstone-ks 

> but actually I think that if the answer to how to manage scopes is "do it via an interceptor," then really both DefaultServiceProvider & CurrentServiceProvider need to be fully public.

I'm not convinced that `DefaultServiceProvider` would need to be public, but I suspect this is something that can come out in the wash! For example if CSLA provides an interceptor implementation for this, it can have internal access to the DefaultServiceProvider (which was given to CSLA on application startup). If you were writing a custom interceptor you could potnetially construct it yourself passing in your root IServiceProvider

```
 public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
 {

   var customInterceptor = new MyCustomInterceptor(app.ApplicationServices); // your apps root sp.
   // however you configure CSLA to use it..
   SetCslaCustomInterceptor(customInterceptor);
}
```

**Comment by ajohnstone-ks** [comment]
@dazinator The interceptor is set by giving it a `System.Type`, and Csla creates the actual instance since there's a new interceptor instance created per DataPortal call, via `Activator.CreateInstance` (I believe), so that suggestion doesn't work, which is why I have that static `SetScopeFactory` method in the interceptor I posted above.

**Comment by ajohnstone-ks** [comment]
At any rate, I think those other changes should probably be their own issue; I think we all agree `ServiceProviderScope` is not a great name for the CurrentServiceProvider.

**Comment by rockfordlhotka** [comment]
I'm happy to change the name in CSLA 6 (the current master branch). I think there's an open issue for this?

One solution for creating these interceptors is that I can create an `InterceptorBase` type that _does_ have a protected reference to the root scope provider maintained by CSLA.

I really don't want to expose the provider to the public in any way, because that opens the door to all sorts of abuse, and I don't want to be responsible for people making poor choices 😀

**Comment by dazinator** [comment]
@rockfordlhotka 

> The interceptor is set by giving it a System.Type, and Csla creates the actual instance since there's a new interceptor instance created per DataPortal call, via Activator.CreateInstance (I believe), so that suggestion doesn't work

I think one way to solve this would be, we could replace the `Activator.CreateInstance` [with this](https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.dependencyinjection.activatorutilities.createinstance?view=dotnet-plat-ext-5.0#Microsoft_Extensions_DependencyInjection_ActivatorUtilities_CreateInstance_System_IServiceProvider_System_Type_System_Object___) and CSLA would use its internal reference to the `DefaultServiceProvider` to activate this type.

In this scenario, users still wouldn't register the type for DI, but CSLA would activate this type using the applications DI container to supply dependencies. So users could register dependencies for this type in their app startup. `IServiceProvider` is supported for injection out of the box, so this means its no longer CSLA's problem to supply this, or any other dependencies to the users custom type.

**Comment by rockfordlhotka** [comment]
I agree - this is part of the plan for #1738 - require that DI be initialized before using CSLA, so then CSLA can rely on DI existing when creating all its types.

**Comment by ajohnstone-ks** [comment]
> I'm happy to change the name in CSLA 6 (the current master branch). I think there's an open issue for this?

That's what this issue is for; it got derailed.
 
> One solution for creating these interceptors is that I can create an `InterceptorBase` type that _does_ have a protected reference to the root scope provider maintained by CSLA.

That would be fine, but I really need both the DefaultSP and the CurrentSP (currently name ServiceProviderScope).  The DefaultSP to call `GetRequiredService<IServiceScopeFactory>` and CurrentSP to check to see if there is a value or not and set it/dispose of it.  I don't want to be creating grand-child containers.

This is also nice because I don't need to look at ApplicationContext at all, so I think that makes things a bit cleaner.

> I really don't want to expose the provider to the public in any way, because that opens the door to all sorts of abuse, and I don't want to be responsible for people making poor choices 😀

I had thought that was the reason for the restrictive getters on those properties, and I understand it.  As long as I can get at the default & current SPs in an interceptor without jumping through a bunch of hoops, that's all I'm looking for.

**Comment by rockfordlhotka** [comment]
fwiw, `master` is now CSLA 6, so a PR to rename the property is a fine thing 😄

**Comment by ajj7060** [comment]
@rockfordlhotka I can take this up tonight if you assign it to me, assuming CurrentServiceProvider is an acceptable name?  If not, let me know what it should be.

**Comment by rockfordlhotka** [comment]
I think that is a good name, yes. This change needs to be only in CSLA 6 (master), because it is a breaking change.

**Comment by ajj7060** [comment]
@rockfordlhotka The current provider is currently stored in LocalContext using a key of `__sps`; should I rename the key as well or just leave it be?  If I rename it, is there any objection to creating a `const` to hold the key name in case it needs to be changed in the future?

**Comment by ajj7060** [comment]
@rockfordlhotka @dazinator 

`CslaConfiguration.ServiceProviderScope(IServiceProvider scope)` will currently set what is now being called CurrentServiceProvider.   I think this is called during application, so I'm thinking it should be renamed to SetDefaultServiceProvider, and actually set ApplicationContext.DefaultServiceProvider instead of CurrentServiceProvider.

What do you think?  I'm also not sure what ramifications this will have, so maybe another issue should be used for that change if it is desired?  LMK.

**Comment by dazinator** [comment]
Yeah please don't change anything as part of this except the renaming.. otherwise from experience it can cause a tonne of issues.

**Comment by dazinator** [comment]
For context this is the basic flow:

1. After application has started up and built its root IServiceProvider - Csla needs to capture a reference to this - this is referred to as the DefaultServiceProvider because its the root application level scope which would be used by default if there is no more specific scope set in csla context.
2. If an application wants CSLA to use dependencies from a different scope the application can build a new scope and give Csla that scoped IServiceProvider. Csla will use this if it has been set, otherwise csla will fall back to using the default service provider as mentioned above.

So `CurrentServiceProvider` should be in reference to point 2. With that in mind I don't think it should be setting DefaultServiceProvider when this is set, as DefaultServiceProvider should only be set once after application startup.

**Comment by ajj7060** [comment]
> So `CurrentServiceProvider` should be in reference to point 2. With that in mind I don't think it should be setting `DefaultServiceProvider`. `DefaultServiceProvider` should only be set once after application startup.

Right, I'm talking about `**CslaConfiguration**.ServiceProviderScope(IServiceProvider scope)`, which I assume is something that should only be called once during application startup.  Right now on startup, the only thing you can do with Service Provider configuration via `CslaConfiguration` is set the current service provider; it seems like via configuration you'd want to set the Default on there instead.  Right now it looks like the only way to set `DefaultServiceProvider` is via the property on `ApplicationContext`.

**Comment by dazinator** [comment]
@ajj7060 oh right, I see!
Yeah I think `CslaConfiguration` should be capturing the `DefaultServiceProvider` so it's probably best to rename that method as well to make that clear

**Comment by ajj7060** [comment]
@rockfordlhotka Can I still add changes against this ticket since its closed?

**Comment by rockfordlhotka** [comment]
I reopened it. What's up?

**Comment by ajj7060** [comment]
@rockfordlhotka 

Thanks.  

I had noted earlier when I renamed `Application.ServiceProviderScope` to `CurrentServiceProvider` that `CslaConfiguration.ServiceProviderScope(scope)` sets `CurrentServiceProvider` which doesn't make sense from an application configuration during startup POV.

So I'd like to rename `CslaConfiguration.ServiceProviderScope(scope)` to `DefaultServiceProvider(provider)` and then also change the method to set `ApplicationContext.DefaultServiceProvider` instead of `ApplicationContext.CurrentServiceProvider`.

Unless you'd rather that stuff be a new issue, but I thought since I discovered it as part of the renaming we could just expand the scope (haha) of this ticket a bit and resolve this.
